package/upgrade/lib.sh

```diff
@@ -317,3 +317,135 @@ shutdown_all_vms()
+  local tmp_file=/tmp/$release_name.tmp.yaml
+  local EXIT_CODE=0
+
+  helm history $release_name -n $namespace -o yaml > $tmp_file || EXIT_CODE=$?
+  if [ $EXIT_CODE != 0 ]; then
+    echo "helm history $release_name -n $namespace failed $EXIT_CODE, try later"
+    return
+  fi
+
+  echo "the current helm history is"
+  helm history $release_name -n $namespace || true
+
+  local cnt=$(yq e '. | length' $tmp_file)
+  if [ "$cnt" -gt 0 ]; then
+    local current_status=$(yq e '.[-1].status' $tmp_file)
+    if [ "$current_status" == "pending-upgrade" ]; then
+
+      # in this state at least 180 seconds, to avoid clash with the normal pending upgrade
+      local dt1=$(yq e '.[-1].updated' $tmp_file)
+      dt1=$(date -d "$dt1" +%s)
+      local dt2=$(date +%s)
+      local delta=$(( dt2 - dt1 ))
+      if [ ! $delta -gt 180 ]; then
+        echo "the release is in pending-upgrade state for $delta seconds, do not rollback"
+        rm -f $tmp_file
+        return
+      fi
+
+      if [ "$cnt" -gt 1 ]; then
+        local previous_status=$(yq e '.[-2].status' $tmp_file)
+        if [ "$previous_status" == "deployed" ]; then
+          local previous_revision=$(yq e '.[-2].revision' $tmp_file)
+          echo "rollback pending release from"
+          yq e '.[-1]' $tmp_file
+          echo "to"
+          yq e '.[-2]' $tmp_file
+
+          helm rollback -n $namespace $release_name $previous_revison || EXIT_CODE=$?
+          if [ $EXIT_CODE != 0 ]; then
+            echo "helm rollback $release_name -n $namespace failed $EXIT_CODE, try later"
+          fi
+
+          echo "check release history after rollback"
+          helm history $release_name -n $namespace || true
+        fi
+      else
+        echo "the $release_name is pending-upgrade, but there is no previous revision to rollback"
+      fi
+    else
+      echo "the $release_name is $current_status"
+      echo "when the helm release has been rolled back, the fleet bundle takes time to be ready, wait"
+    fi
+  else
+    echo "there is no helm history of $release_name in $namespace"
+  fi
+  rm -f $tmp_file
+}
+
+
+check_fleet_bundle_helm_releases()
+{
+  local tmp_file=/tmp/bundle_status.tmp.txt
+  local EXIT_CODE=0
+  # the bundle object is too big, fetch the wanted information only
+  kubectl get bundle -A -o=custom-columns='NAME:.metadata.name,HELM_NAME:.spec.helm.releaseName,NAMESPACE:.spec.defaultNamespace,MESSAGE:.status.conditions[?(@.type=="Ready")].message' --no-headers > $tmp_file || EXIT_CODE=$?
+
+  if [ $EXIT_CODE != 0 ]; then
+    echo "kubectl get bundle failed $EXIT_CODE, try later"
+    return
+  fi
+
+  while read -r name helm_name ns message; do
+    if [[ $message == *"another operation"* ]]; then
+      echo "bundle $name is pending_upgrade: $message"
+      if [[ $helm_name == "<none>" ]]; then
+        rname=$name
+      else
+        rname=$helm_name
+      fi
+      echo "prepare to rollback $rname in $ns"
+      try_rollback_helm_release $ns $rname
```

Comment:
@w13915984028 I suggest we don't do automatic rollback here. The previous revision is unknown, we don't know the consequence to downgrade a chart.

Reply:
now the automatical rollback is controlled by an environment, and it will NOT trigger but default

it waits upto 3 mintues and leave it there:
```
Thu 16 Mar 2023 12:44:24 PM CET
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1                       
fleet-local   local-managed-system-agent                    1/1                       
fleet-local   mcc-harvester                                 1/1                       
fleet-local   mcc-harvester-crd                             1/1                       
fleet-local   mcc-local-managed-system-upgrade-controller   1/1                       
fleet-local   mcc-rancher-logging                           1/1                       
fleet-local   mcc-rancher-logging-crd                       1/1                       
fleet-local   mcc-rancher-monitoring                        0/1                       ErrApplied(1) [Cluster fleet-local/local: another operation (install/upgrade/rollback) is in progress]
fleet-local   mcc-rancher-monitoring-crd                    1/1                       
some bundles are not ready:True True True True True True True False True

```
....
```
Thu 16 Mar 2023 12:47:24 PM CET
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1                       
fleet-local   local-managed-system-agent                    1/1                       
fleet-local   mcc-harvester                                 1/1                       
fleet-local   mcc-harvester-crd                             1/1                       
fleet-local   mcc-local-managed-system-upgrade-controller   1/1                       
fleet-local   mcc-rancher-logging                           1/1                       
fleet-local   mcc-rancher-logging-crd                       1/1                       
fleet-local   mcc-rancher-monitoring                        0/1                       ErrApplied(1) [Cluster fleet-local/local: another operation (install/upgrade/rollback) is in progress]
fleet-local   mcc-rancher-monitoring-crd                    1/1                       
some bundles are not ready:True True True True True True True False True
finish wait

```